### PR TITLE
Refactor two phase flow problem

### PR DESCRIPTION
**Pull request by mschreter**
references #211 and #221; ~depends on #251~

~I'll rebase once #251 is merged.~ Then I am ready to be reviewed :-). @peterrum @nmuch FYI

This PR provides the following refactoring of the `TwoPhaseFlowProblem`/`MeltPoolOperation`:

- eliminate problem types: `two_phase_flow_with_heat_transfer|two_phase_flow_with_evaporation|melt_pool_with_evaporation`
- introduce flags: `do_heat_transfer|do_evaporation|do_melt_pool`
- remove `HeatTransferOperation` from `MeltPoolOperation` --> use the `HeatTransferOperation` from `TwoPhaseFlowProblem`
- split up `MeltPoolOperation::solve()` into `MeltPoolOperation::compute_heat_source()`, `MeltPoolOperation::compute_melt_front_propagation()` and `MeltPoolOperation::compute_force_flow_rhs()`

In a follow-up PR I will:
- rename the `TwoPhaseFlowProblem` to the `MeltPoolProblem`
- simultaneously eliminate the problem type `two_phase_flow`
- move the parameters from `flow` to the corresponding section `melt pool`

### CODE REVIEW

**Review by peterrum — APPROVED** [review]
I like the changes of this PR and the current state of the code! The algorithm in `TwoPhaseFlowProblem::run()` looks amazingly clean now and is easy to understand!

I guess it is time to say goodbye to `TwoPhaseFlowProblem`!

**Comment by peterrum** [comment]
TravisCI has not reappeared yet :(

File: source/heat/laser_analytical_temperature_field.cpp [hunk 1]
```diff
@@ -48,8 +48,8 @@ namespace MeltPoolDG::Heat
                                               laser_position);
         }
 
-    temperature.compress(VectorOperation::insert);
-    scratch_data.get_constraint(temp_dof_idx).distribute(temperature);
+    // temperature.compress(VectorOperation::insert);
+    // scratch_data.get_constraint(temp_dof_idx).distribute(temperature);
```
[the remark this answers is not in the document]
Comment (mschreter) [reply source/heat/laser_analytical_temperature_field.cpp:52]: This change was not on purpose, thanks for pointing it out.

**Comment by mschreter** [comment]
Thanks @peterrum for the support in this case :-). I've applied the changes. From my side this PR can be merged any time :-).

**Review by peterrum — APPROVED** [review]

File: tests/melt_pool_tfi_static_interface_gauss_no_solidification.with_adaflo_support=on.mpirun=4.output [hunk 2]
```diff
@@ -1,41 +1,41 @@
 current laser position: 0 0
 current laser intensity: 1
- 	 |n_0| = 5.2501595611e-14  |n_1| = 4.4301244095    
-    | reini: τ= 4.441875e-06 |Ψ RHS|² = 2.65265638872339e-16 |
+ 	 |n_0| = 4.6334517697e-14  |n_1| = 4.4301244095    
+    | reini: τ= 4.441875e-06 |Ψ RHS|² = 2.65265649234409e-16 |
```
Comment (nmuch) [inline tests/melt_pool_tfi_static_interface_gauss_no_solidification.with_adaflo_support=on.mpirun=4.output:4]: Why did the values in this test case change so much?

**Comment by peterrum** [comment]
CI is not happy:

> The following tests FAILED:
	 15 - tests/recoil_pressure.mpirun=4.debug (Failed)
	 16 - tests/recoil_pressure_fixed_solid_domain.mpirun=4.debug (Failed)
	 17 - tests/recoil_pressure_simplex.mpirun=4.debug (Failed)

The stack trace for one of them is:
```
17: --------------------------------------------------------
17: An error occurred in line <624> of file <../include/deal.II/base/partitioner.templates.h> in function
17:     void dealii::Utilities::MPI::Partitioner::import_from_ghosted_array_finish(dealii::VectorOperation::values, const dealii::ArrayView<const ElementType, MemorySpaceType>&, const dealii::ArrayView<ElementType, MemorySpace>&, const dealii::ArrayView<ElementType, MemorySpace>&, std::vector<ompi_request_t*, std::allocator<ompi_request_t*> >&) const [with Number = double; MemorySpaceType = dealii::MemorySpace::Host]
17: The violated condition was: 
17:     *read_position == Number() || internal::get_abs(locally_owned_array[j] - *read_position) <= internal::get_abs(locally_owned_array[j] + *read_position) * 100000. * std::numeric_limits<typename numbers::NumberTraits< Number>::real_type>::epsilon()
17: Additional information: 
17:     Called compress(VectorOperation::insert), but the element received
17:     from a remote processor, value 1, does not match with the value 0 on
17:     the owner processor 0
17: 
17: Stacktrace:
17: -----------
17: #0  /usr/local/lib/libdeal_II.g.so.10.0.0-pre: void dealii::Utilities::MPI::Partitioner::import_from_ghosted_array_finish<double, dealii::MemorySpace::Host>(dealii::VectorOperation::values, dealii::ArrayView<double const, dealii::MemorySpace::Host> const&, dealii::ArrayView<double, dealii::MemorySpace::Host> const&, dealii::ArrayView<double, dealii::MemorySpace::Host> const&, std::vector<ompi_request_t*, std::allocator<ompi_request_t*> >&) const
17: #1  /usr/local/lib/libdeal_II.g.so.10.0.0-pre: dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>::compress_finish(dealii::VectorOperation::values)
17: #2  /usr/local/lib/libdeal_II.g.so.10.0.0-pre: dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>::compress(dealii::VectorOperation::values)
17: #3  /home/dealii/adaflo-build/MeltPoolDG-build/meltpooldg: MeltPoolDG::MeltPool::MeltPoolOperation<2>::compute_solid_and_liquid_phases(dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host> const&)
17: #4  /home/dealii/adaflo-build/MeltPoolDG-build/meltpooldg: MeltPoolDG::MeltPool::MeltPoolOperation<2>::set_initial_condition(dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host> const&, dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>&)
17: #5  /home/dealii/adaflo-build/MeltPoolDG-build/meltpooldg: MeltPoolDG::Flow::TwoPhaseFlowProblem<2>::set_initial_condition(std::shared_ptr<MeltPoolDG::SimulationBase<2, 2> >)
17: #6  /home/dealii/adaflo-build/MeltPoolDG-build/meltpooldg: MeltPoolDG::Flow::TwoPhaseFlowProblem<2>::initialize(std::shared_ptr<MeltPoolDG::SimulationBase<2, 2> >)
17: #7  /home/dealii/adaflo-build/MeltPoolDG-build/meltpooldg: MeltPoolDG::Flow::TwoPhaseFlowProblem<2>::run(std::shared_ptr<MeltPoolDG::SimulationBase<2, 2> >)
17: #8  /home/dealii/adaflo-build/MeltPoolDG-build/meltpooldg: void MeltPoolDG::Simulation::run_simulation<double>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, ompi_communicator_t*)
17: #9  /home/dealii/adaflo-build/MeltPoolDG-build/meltpooldg: main
17: --------------------------------------------------------
```

Magdalena, do you an idea which one of your changes in the morning could have caused this? I can reproduce the problem locally on my machine.

**Review by nmuch — APPROVED** [review]
Thank you, @mschreter, for the refactoring! I think the two phase flow problem is a lot more readable now. I'll rebase #252 after this is merged.

File: tests/melt_pool_tfi_static_interface_gauss_no_solidification.with_adaflo_support=on.mpirun=4.output (hunk 2 above)
Comment (peterrum) [reply tests/melt_pool_tfi_static_interface_gauss_no_solidification.with_adaflo_support=on.mpirun=4.output:4]: The file was created on your machine so slight differences are to be expected. The absolute difference is around machine epsilon.

**Comment by mschreter** [comment]
> CI is not happy:
> 
> > The following tests FAILED:
> > 15 - tests/recoil_pressure.mpirun=4.debug (Failed)
> > 16 - tests/recoil_pressure_fixed_solid_domain.mpirun=4.debug (Failed)
> > 17 - tests/recoil_pressure_simplex.mpirun=4.debug (Failed)
> 
> The stack trace for one of them is:
> 
> ```
> 17: --------------------------------------------------------
> 17: An error occurred in line <624> of file <../include/deal.II/base/partitioner.templates.h> in function
> 17:     void dealii::Utilities::MPI::Partitioner::import_from_ghosted_array_finish(dealii::VectorOperation::values, const dealii::ArrayView<const ElementType, MemorySpaceType>&, const dealii::ArrayView<ElementType, MemorySpace>&, const dealii::ArrayView<ElementType, MemorySpace>&, std::vector<ompi_request_t*, std::allocator<ompi_request_t*> >&) const [with Number = double; MemorySpaceType = dealii::MemorySpace::Host]
> 17: The violated condition was: 
> 17:     *read_position == Number() || internal::get_abs(locally_owned_array[j] - *read_position) <= internal::get_abs(locally_owned_array[j] + *read_position) * 100000. * std::numeric_limits<typename numbers::NumberTraits< Number>::real_type>::epsilon()
> 17: Additional information: 
> 17:     Called compress(VectorOperation::insert), but the element received
> 17:     from a remote processor, value 1, does not match with the value 0 on
> 17:     the owner processor 0
> 17: 
> 17: Stacktrace:
> 17: -----------
> 17: #0  /usr/local/lib/libdeal_II.g.so.10.0.0-pre: void dealii::Utilities::MPI::Partitioner::import_from_ghosted_array_finish<double, dealii::MemorySpace::Host>(dealii::VectorOperation::values, dealii::ArrayView<double const, dealii::MemorySpace::Host> const&, dealii::ArrayView<double, dealii::MemorySpace::Host> const&, dealii::ArrayView<double, dealii::MemorySpace::Host> const&, std::vector<ompi_request_t*, std::allocator<ompi_request_t*> >&) const
> 17: #1  /usr/local/lib/libdeal_II.g.so.10.0.0-pre: dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>::compress_finish(dealii::VectorOperation::values)
> 17: #2  /usr/local/lib/libdeal_II.g.so.10.0.0-pre: dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>::compress(dealii::VectorOperation::values)
> 17: #3  /home/dealii/adaflo-build/MeltPoolDG-build/meltpooldg: MeltPoolDG::MeltPool::MeltPoolOperation<2>::compute_solid_and_liquid_phases(dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host> const&)
> 17: #4  /home/dealii/adaflo-build/MeltPoolDG-build/meltpooldg: MeltPoolDG::MeltPool::MeltPoolOperation<2>::set_initial_condition(dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host> const&, dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>&)
> 17: #5  /home/dealii/adaflo-build/MeltPoolDG-build/meltpooldg: MeltPoolDG::Flow::TwoPhaseFlowProblem<2>::set_initial_condition(std::shared_ptr<MeltPoolDG::SimulationBase<2, 2> >)
> 17: #6  /home/dealii/adaflo-build/MeltPoolDG-build/meltpooldg: MeltPoolDG::Flow::TwoPhaseFlowProblem<2>::initialize(std::shared_ptr<MeltPoolDG::SimulationBase<2, 2> >)
> 17: #7  /home/dealii/adaflo-build/MeltPoolDG-build/meltpooldg: MeltPoolDG::Flow::TwoPhaseFlowProblem<2>::run(std::shared_ptr<MeltPoolDG::SimulationBase<2, 2> >)
> 17: #8  /home/dealii/adaflo-build/MeltPoolDG-build/meltpooldg: void MeltPoolDG::Simulation::run_simulation<double>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, ompi_communicator_t*)
> 17: #9  /home/dealii/adaflo-build/MeltPoolDG-build/meltpooldg: main
> 17: --------------------------------------------------------
> ```
> 
> Magdalena, do you an idea which one of your changes in the morning could have caused this? I can reproduce the problem locally on my machine.

I ran the tests only in release mode :/. Yes, the problem is the `VectorOperation::insert`. I'll replace this by `VectorOperation::max` which should be fine.

**Comment by mschreter** [comment]
One missing `update_ghost_values()` and you are lost :-).

File: source/melt_pool/melt_pool_operation.cpp [hunk 3]
```diff
@@ -270,6 +270,7 @@ namespace MeltPoolDG::MeltPool
   MeltPoolOperation<dim>::compute_solid_and_liquid_phases(const VectorType &level_set_as_heaviside)
   {
     level_set_as_heaviside.update_ghost_values();
+    temperature->update_ghost_values();
```
Comment (mschreter) [inline source/melt_pool/melt_pool_operation.cpp:273]: One missing `update_ghost_values()` and you are lost ;-).

**Review by peterrum — APPROVED** [review]
Odd. Shouldn't an assert be risen somewhere else? I guess the vector was still ghosted but with the old values!?

**Comment by mschreter** [comment]
> Odd. Shouldn't an assert be risen somewhere else? I guess the vector was still ghosted but with the old values!?

In this case not because I access the entries by index:

```cpp
solid[local_dof_indices[i]] =  is_solid_region(level_set_as_heaviside[local_dof_indices[i]], 
                                                                      (*temperature)[local_dof_indices[i]]);
```

**Comment by peterrum** [comment]
> In this case not because I access the entries by index:

This does not mapper. In any case, an assert should have been risen. Probably a `zero_out_ghost_values()` was missing.

**Pull request merged**